packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1271 @@
+   */
+  public metricSearchLatency(props?: MetricOptions): Metric {
+    return this.metric('SearchLatencyP99', { statistic: 'p99', ...props });
+  }
+
+  /**
+   * Metric for indexing latency.
+   *
+   * @default p99 over 5 minutes
+   */
+  public metricIndexingLatency(props?: MetricOptions): Metric {
+    return this.metric('IndexingLatencyP99', { statistic: 'p99', ...props });
+  }
+
+  private grant(
+    grantee: iam.IGrantable,
+    domainActions: string[],
+    resourceArn: string,
+    ...otherResourceArns: string[]
+  ): iam.Grant {
+    const resourceArns = [resourceArn, ...otherResourceArns];
+
+    const grant = iam.Grant.addToPrincipal({
+      grantee,
+      actions: domainActions,
+      resourceArns,
+      scope: this,
+    });
+
+    return grant;
+  }
+}
+
+
+/**
+ * Reference to an Elasticsearch domain.
+ */
+export interface DomainAttributes {
+  /**
+   * The ARN of the Elasticsearch domain.
+   */
+  readonly domainArn: string;
+
+  /**
+   * The domain name of the Elasticsearch domain.
+   */
+  readonly domainName: string;
+
+  /**
+   * The domain endpoint of the Elasticsearch domain.
+   */
+  readonly domainEndpoint: string;
+}
+
+
+/**
+ * Provides an Elasticsearch domain.
+ */
+export class Domain extends DomainBase implements IDomain {
+  /**
+   * Creates a Domain construct that represents an external domain via domain endpoint.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param domainEndpoint The domain's endpoint.
+   */
+  public static fromDomainEndpoint(
+    scope: cdk.Construct,
+    id: string,
+    domainEndpoint: string,
+  ): IDomain {
+    const stack = cdk.Stack.of(scope);
+    const domainName = extractNameFromEndpoint(domainEndpoint);
+    const domainArn = stack.formatArn({
+      service: 'es',
+      resource: 'domain',
+      resourceName: domainName,
+    });
+
+    return Domain.fromDomainAttributes(scope, id, {
+      domainArn,
+      domainName,
+      domainEndpoint,
+    });
+  }
+
+  /**
+   * Creates a Domain construct that represents an external domain.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param attrs A `DomainAttributes` object.
+   */
+  public static fromDomainAttributes(scope: cdk.Construct, id: string, attrs: DomainAttributes): IDomain {
+    return new class extends DomainBase {
+      public readonly domainArn = attrs.domainArn;
+      public readonly domainName = attrs.domainName;
+      public readonly domainEndpoint = attrs.domainEndpoint;
+
+      constructor() { super(scope, id); }
+    };
+  }
+
+  public readonly domainArn: string;
+  public readonly domainName: string;
+  public readonly domainEndpoint: string;
+
+
+  private readonly domain: CfnDomain;
+  private readonly slowSearchLogGroup?: logs.ILogGroup;
+  private readonly slowIndexLogGroup?: logs.ILogGroup;
+  private readonly appLogGroup?: logs.ILogGroup;
```

Comment:
These should be public as they can be implicitly created on the user's behalf, so he/she would need access to it.